[Team-19][AOS] [ Linus& 히데 Airbnb 첫번째주 금요일 PR ] 달력 화면 일부 구현

cherrylime69 opened the pull request:
## 진행 사항
### 커스텀 달력 UI 구현
- Calendar 프래그먼트 생성
- 리사이클러뷰를 중첩하여 UI 구현 계획
- 각 월, 일에 대한 달력 정보를 생성 및 처리하기 위해 CalendarUtil 정의
- CalendarUtil 를 통해 가공된 데이터를 월, 일 어댑터에 주어 리사이클러뷰로 UI는 구현 완료
### 달력 데이터 처리 로직 수정
- 달력의 각 일 정보 처리를 위한 CalendarDay 클래스 생성
- 달력의 각 월별 데이터 생성을 Repository에서 생성하는 방식 테스트 (데이터 생성 및 처리를 Repository에 옮기는 중)
- 현재 날짜 기준 이전 날짜에 대한 UI 처리 로직 추가
- 현재 날짜에 대한 UI 처리 로직 추가 (현재 날짜면 초록 배경색으로)

### 결과 화면
![res](https://user-images.githubusercontent.com/95393311/170625379-3023ed74-c78d-4529-b038-39d1926bf64b.gif)

---

### 참고 사항
- 현재 레포지터리는 테스트를 진행하고 있어 리뷰하지 않으셔도 됩니다.
- 현재 1년치 월, 일에 대한 데이터를 레포지터리에서 생성 및 관리하도록 수정하고 있습니다. (이전에는 어댑터에서 데이터를 생성했었음)

### 질문 사항
1. 캘린더 리사이클러뷰에 넣을 데이터 리스트를 레포지토리에서 생성하고 viewModel- view를 통해 UI 업데이트를 하는 방식으로 수정하려고 하고 있는데 적절한 방법인지 궁금합니다.

![앱바](https://user-images.githubusercontent.com/95393311/170626888-8427321e-8cbd-4fd5-b45c-2d56eb231cd1.JPG)
2. 위의 이미지를 보시면 달력 화면 - 요금 입력 화면 - 인원 입력 화면에서 상단 UI 가 똑같아서 처음에는 앱바로 구현하려고 했습니다. 하지만 앱바의 메뉴를 활용하지도 않고, 클릭했을 때 tint 색을 바꾸는 코드가 너무 복잡해서 앱바가 아니라 imageButton 및 TextView로 구현을 했습니다. 혹시 앨런은 이 UI 에서 앱바를 사용하는 게 더 적절하다고 생각하시나요?

kang-proto committed:
feat: 정보 입력 페이지 UI 완성

- 상단 앱바 부분 완성
- 하단 앱바 부분 완성
- 정보 입력 타입마다 fragment 변경 되도록 container+ navgraph 추가

cherrylime69 committed:
Merge pull request #20 from hwicode/feature/informationFragment

[Android][히데] 정보 입력 페이지 UI 완성

cherrylime69 committed:
feat : 커스텀 달력 일부 구현

- Calendar 프래그먼트 생성
- 리사이클러뷰를 중첩하여 간단한 UI 구현 완료

kang-proto committed:
feat: 달력 UI 완성
- 연,월 정보 처리를 위한 리사이클러뷰 구현
- 각 월에 대한 달력정보를 처리하기 위해 CalendarUtil 정의
- 각 월에 대한 달력정보를 리사이클러뷰로 구현

kang-proto committed:
feat: 달력 데이터 처리 로직 수정

- 달력의 각 일 정보 처리를 위한 CalendarDay 클래스 생성
- 달력의 각 월별 데이터 생성을 Repository에서 생성하는 방식 테스트
- 현재 날짜 기준 이전 날짜에 대한 UI 처리 로직 추가
- 현재 날짜에 대한 UI 처리 로직 추가

cherrylime69 committed:
Merge pull request #26 from hwicode/feature/custom-calendar

[Android][히데_라이너스] 달력 화면 기능 구현 진행

cherrylime69 committed:
Merge pull request #27 from hwicode/android

[Android] 금요일 두번째 PR

kang-proto committed:
hotfix: codeFormatting, UI 수정

sabgilhun commented on app/build.gradle:
```diff
@@ -56,4 +60,7 @@ dependencies {
     kapt "com.github.bumptech.glide:compiler:4.13.0"
+
+    //Joda
+    implementation 'joda-time:joda-time:2.10.10'
```
`joda-time`을 사용하게 된 이유를 알고 싶습니다!

어떤 장점을 기대하고 선택하게 되었는지 말씀해주실 수 있을까요?

그냥 의견 나누고 싶은거라 편하게 말씀해주셔도 됩니다~

HyungMinKang replied:
Date,Calendar클래스의 경우 날짜를 표시하는데  불변객체가 아니고 월 표시가 모호해서 부적절하다 생각했습니다.   그래서 LocalDateTime을 사용하려고 했는데,   오레오  버전 이하에 대해 분기문 처리를 해줘야 하는 부분이 있어,  분기문 처리 없이 한번에 해결할 수 있는 joda 라이브러리를 사용하게 되었습니다.

sabgilhun replied:
@HyungMinKang 

오... 저도 그렇게 생각합니다. 

Java 8에서 Date, Calendar의 문제들을 해결한 JSR310이 추가 되었지만, Android에선 모든 Java 8 기능들이 포함되지 않았습니다.

하지만 최근에 '자바 8+ API 디슈가링'을 지원하게 되면서 미포함됐던, Java 8 기능이 추가로 더 사용 가능해지게 되었습니다.
(https://developer.android.com/studio/write/java8-support?hl=ko#library-desugaring)
(https://developer.android.com/studio/write/java8-support-table?hl=ko)

사용하시려는 부분들도 해당 목록에 있을 수 있어요. 한번 확인 해보셔도 좋을것 같네요.

또한, joda time 라이브러리 공식 웹페이지에 가면 아래와 같이 적혀 있는데요.

> The standard date and time classes prior to Java SE 8 are poor. By tackling this problem head-on, Joda-Time became the de facto standard date and time library for Java prior to Java SE 8. Note that from Java SE 8 onwards, users are asked to migrate to java.time (JSR-310) - a core part of the JDK which replaces this project.

Java SE 8 이전의 time 클래스의 문제를 해결하기 위한 라이브러리이며, JSR-310이 포함된 Java SE 8 이후 부턴 'java.time'으로 마이그레이션을 권장하고 있습니다.

그렇기에 최대한 사용하지 않는것이 좋아보입니다. 'java.time'을 사용할 수 있는지 확인 해보는 것이 좋겠네요.

sabgilhun commented on app/src/main/java/com/example/airbnb/data/RepositoryImpl.kt:
```diff
@@ -0,0 +1,18 @@
+    override fun getCalendarMap(): Map<LocalDateTime, List<CalendarDay>> {
+        var calendarMap: MutableMap<LocalDateTime, List<CalendarDay>> = mutableMapOf()
+        val monthList = Array(12) { index -> LocalDateTime.now().plusMonths(index) }
+        monthList.map {
```
`map`
https://kotlinlang.org/docs/collection-transformations.html

`forEach`
https://kotlinlang.org/docs/iterators.html

`map`과 `forEach`는 용도가 다른 함수인데요. 지금의 코드는 `forEach`이 더 어울릴 것 같네요.

동일한 결과를 내더라도 내부 동작 과 의미 파악에 도움디 되게 적절히 사용하는 것이 좋습니다.

cherrylime69 replied:
map - 반환값 o, forEach - 반환값 x 인 점을 학습하였습니다. 피드백대로 해당 코드는 반환값이 필요없으니 forEach 로 대체하였습니다.

sabgilhun commented on app/src/main/java/com/example/airbnb/ui/calendar/DayAdapter.kt:
```diff
@@ -0,0 +1,45 @@
+    class ViewHolder(private val binding: ItemCalendarDayBinding) : RecyclerView.ViewHolder(binding.root) {
+        fun bind(day: CalendarDay) {
+            binding.day = day.day
+            if (day.isSelectable) {
+                if (day.isStartDay) {
+                    binding.tvCalendarDay.setBackgroundColor(Color.GREEN)
+                }
+                binding.tvCalendarDay.setTextColor(Color.BLACK)
+            } else {
+                binding.tvCalendarDay.setTextColor(Color.GRAY)
+            }
```
이미 알고 계실지도 모르지만,

특정 뷰에 setXXXX로 되어 있는 메서드 전부 databinding이 가능합니다.

https://developer.android.com/topic/libraries/data-binding/binding-adapters?hl=ko#automatic-setting

```
// kotlin
View.setBackgroundColor(..) // 뷰에 해당 메서드가 존재하면 아래와 같이 사용 가능

// xml
<TextView
    android:id="@+id/tv_calendar_day"
    ...
    android:backgroundColor="@{Color.GREEN}"
    ...
/>
```

지금의 경우 view 설정을 '`bind`에서 하냐', '`xml`에서 하냐'는 개인의 취향이라고 생각하기에 수정하지 않으셔도 됩니다.

cherrylime69 replied:
오, 처음으로 알게 되었습니다. 좋은 가이드 감사합니다! 처음 본거라 bind 에서 백그라운드 색상을 바꾸도록 변경하였습니다.

sabgilhun commented on app/src/main/java/com/example/airbnb/ui/calendar/MonthAdapter.kt:
```diff
@@ -0,0 +1,41 @@
+
+class MonthAdapter : RecyclerView.Adapter<MonthAdapter.ViewHolder>() {
+
+    private val _monthList = Array(12) { index ->
+        LocalDateTime.now().plusMonths(index)
+    }
```
List 대신 Array를 사용한 이유가 무엇인가요? 🤔

단순히 의견을 듣고 싶어 남긴 코멘트라 편히 답변주세요~

HyungMinKang replied:
고정된 사이즈이고  컬렉션이 수정이나 변경이 일어나지 않아서  Array를 써도 괜찮다고 생각했습니다.

sabgilhun commented on app/src/main/java/com/example/airbnb/ui/common/CalendarUtil.kt:
```diff
@@ -0,0 +1,106 @@
+import com.example.airbnb.domain.model.CalendarDay
+import kotlin.math.absoluteValue
+
+object CalendarUtil {
```
`CalendarUtil ` 내부에서 사용되는 magic number 들이 몇개 있는데요.

```kotlin
if (dayOfWeek != 7) {
...
}
```

``` kotlin
for (i in 8..49) {
...
}
```

위의 코드처럼 어떤 의도로 쓰였는지 알 수 없이 숫자만 사용된 경우, 상수를 선언하여 가독성을 높힐 수 있습니다!

https://www.codebyamir.com/blog/software-anti-patterns-magic-numbers

cherrylime69 replied:
저희가 미처 그 부분까지는 생각하지 못하였습니다. enum class 를 만들어 해당 숫자의 의미를 명시했습니다

sabgilhun commented on app/src/main/java/com/example/airbnb/ui/common/CalendarUtil.kt:
```diff
@@ -0,0 +1,106 @@
+        month: Int,
+        dayOfWeek: Int,
+        lastDay: Int
+    ): ArrayList<CalendarDay> {
```
구체 타입보단 interface를 리턴해주는 방식을 지향하는 것이 좋습니다.

```kotlin
private fun getDayListStartWithSunday(
    month: Int,
    dayOfWeek: Int,
    lastDay: Int
): List<CalendarDay> {
    ...
}
```

cherrylime69 replied:
List 가 Collect 의 구현체인 점을 새롭게 배웠습니다. ArrayList 와 List 의 계층도를 비교하는데, 명확하게 어떤 차이점이 있어서 말씀하신  `구체 타입보단 interface를 리턴해주는 방식을 지향하는 것이 좋습니다` 에 대한 조언을 이해하지 못했습니다. 혹시 어떤 점 때문에 interface 를 리턴해 주는 방식(여기서는 List 타입을 반환하는 것)이 더 나은 것인지 힌트나 관련된 사이트를 주실 수 있으실까요?

sabgilhun replied:
@cherrylime69 

이런 상황을 상상해보면 좋을것 같습니다.

cherrylime69께서 오픈 소스 라이브러리를 만든다고 생각해봅시다.

그 라이브러리에선 아래와 같은 API를 제공합니다.

``` kotlin
fun getDayList(date: LocalDateTime): ArrayList<CalendarDay> {
    ...       
}
```

그리고 저는 라이브러리의 `getDayList` 함수를 프로젝트 전역에서 많이 사용하고 있었습니다.


``` kotlin
fun printDataList(list: ArrayList<CalendarDay>) {
    ...
}

val dayList = getDayList(date)
printDataList(dayList)
```

그 후 cherrylime69께서 성능 개선을 위해 아래와 같이 오픈 소스 라이브러리를 업데이트 했다고 가정해봅시다.

``` kotlin
fun getDayList(date: LocalDateTime): LinkedList<CalendarDay> {
    ...       
}
```

그럼 라이브러리를 프로젝트 전반에서 사용하던 저는 아래와 같은 코드들을 전부 변경해야합니다.

``` kotlin
// as-is
fun printDataList(list: ArrayList<CalendarDay>) {
    ...
}

// to-be
fun printDataList(list: LinkedList<CalendarDay>) {
    ...
}
```

위와 같은 문제는 `List` 인터페이스를 return 했다면, 문제가 되지 않아요. 이런 방식이 서로의 의존성을 줄일 수 있는 인터페이스의 장점 중 하나라고 생각하시면 됩니다.

여기선 극단적인 상황을 예시로 들기 위해 오픈 라이브러리를 예시로 들었지만, Util 함수도 동일 합니다. `ArrayList`를 꼭 return 해야하는 이유가 없다면 최대한 추상화 타입을 반환하는 것이 좋습니다.

sabgilhun commented on app/src/main/java/com/example/airbnb/ui/common/CalendarUtil.kt:
```diff
@@ -0,0 +1,106 @@
+        val nowDay = org.joda.time.LocalDateTime.now().dayOfMonth
+        for (i in 8..49) {
+            if (i <= dayOfWeek || i > lastDay + dayOfWeek) {
+                dayList.add(CalendarDay(month, ""))
```
```
CalendarDay(month, "")
```

default 값 사용으로 부분적인 인자로 생성자(함수)를 호출하는 경우, 
인자 명을 명시해주는 것이 추후에 실수를 줄일 수 있습니다.

``` kotlin
CalendarDay(month = month, day = "")
```

`CalendarDay`에 만약 추가로 String 인자가 추가 되면,
`day` 인자에 들어갈 값이 새로 추가된 인자에 들어가도 인지하기 쉽진 않기 때문입니다.

cherrylime69 replied:
네, 수정하였습니다

sabgilhun commented on app/src/main/res/navigation/navigation_information.xml:
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<navigation xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:id="@+id/navigation_information"
+    app:startDestination="@id/calendarFragment">
+
+    <fragment
+        android:id="@+id/calendarFragment"
+        android:name="com.example.airbnb.ui.calendar.CalendarFragment"
+        android:label="fragment_calendar"
+        tools:layout="@layout/fragment_calendar" />
+</navigation>
```
검색 필터링 (날짜 입력, 요금 입력, 인원 입력) 등 화면을 Navigation Component 로 다루는 방식 적절하게 사용된 느낌이 드네요 👍 

이런 구조를 잡을때 어떤 샘플같은 걸 보고 결정 하셨을까요? 어떤 판단기준에 의해 구조를 잡게 되었는지 궁금하네요.

HyungMinKang replied:
여러번의 프로젝트를 거치면서 Navigation Componenet를 많이 다루게 되어 익숙해진 영향이 있는 것 같습니다. 특별히 샘플같은것을 참고하기 보다는,  화면 디자인 기획서에서 공통되는 부분이 눈에 띄어서  현재 구조로 구성하게 된 것 같습니다.

sabgilhun reviewed (commented):
안녕하세요. 라이너스, 히데 이번주에 캘린더 부분 작업을 하신것 같네요.

은근히 캘린더 데이터를 다루는 법이 복잡해서 쉽지 않았을텐데, 잘 구현하셨네요. 멋집니다.

## 질문 답변
### Q1. 캘린더 데이터 리스트 repo or adapter ??

Data Layer(Repository) 에서 어떤 작업을 해야할지에 대해 먼저 정리를 하는 것이 좋아보이는데요.

저는 Repository에선 Data를 만드는 작업보단, Data를 가져오는 작업들을 하는 레이어라 생각합니다.

여기 Android 데이터영역 문서 '일반적인 작업' 쪽을 보면, 이해하는데에 조금 도움이 될것 같습니다.
(https://developer.android.com/jetpack/guide/data-layer#common-tasks)

**캘린더 데이터 리스트를 가져오는 과정**을 세분화 해보겠습니다.
1. 현재 날짜를 가져온다. 
2. 현재 날짜 기준으로 1년(12달) Data를 리스트로 만든다.
3. 각 `달`에 존재하는 `날`을 계산하여 매핑한다.

제 기준으론 1번까지가 Repository의 역할로 보입니다. 나머진 Domain or ViewModel에서 하는 방식으로 작업 했을것 같아요.

제 의견을 정리하자면
* 데이터를 레포를 통해 가져오는 것은 OK
* 하지만 레포에서 모든 데이터를 가공해서 전달하는 것은 No

**추가로 현재 날짜는 서버에서 받는것이 좋아요. 모바일 날짜는 조작이 가능해서, 신뢰하면 안됩니다.**

### Q2 검색 필터쪽, AppBar 대신 ImageButton, TextView 조합하여 구현
저는 문제 없다고 생각합니다. `AppBarLayout` 조차도 `LinearLayout`을 확장한 뷰입니다.
대신 `AppBarLayout` 구현되어 있는 여러가지 기능을 쓰지 못할뿐인데, 지금 그런 경우가 없다고 하셨으니 문제 없을것 같습니다. 좋은 판단 기준을 찾기위한 질문 좋네요.

## 추가로 고민해봤으면 하는 부분입니다. (수정하지 않으셔도 됩니다.)
* `CalendarUtil`에서 캘린더 UI에서 빈칸이 노출되는 부분까지 로직까지 포함되어 있는데, 과연 이러한 로직이 Util로 빠질만큼 일반적인 로직일까? 해당 화면에서만 사용될 로직이지 않을까?

kang-proto committed:
fix : 첫번째주 두번째 PR 리뷰 반영
- setBackground -> bind로 변경
- CalendarUtil 매직 넘버 -> Constants에 상수처리
- default값 처리 수정
- 레포지토리에서 캘린더 데이터 생성 코드 삭제 -> viewModel에서 처리하는 방향으로 수정예정

cherrylime69 committed:
Merge pull request #35 from hwicode/fix/firstWeekSecondReview

[Android][히데] fix : 첫번째주 두번째 PR 리뷰 반영

HyungMinKang commented:
리뷰 주신 내용을 바탕으로 수정을 진행하였습니다.  레포지토리에서 캘린더 데이터를 생성하는 부분은 제거를 하였습니다. 날짜 데이터 정보가 UI에서만 사용되어서  별도로 생성 코드를 분리하기보다는  ViewModel에서 처리하는 방식으로 진행하겠습니다.  날짜 정보를 서버에서 받아오는 부분은  백엔드 분들과 타협해서 저희가 처리하는 방향으로 결정했습니다. ( 다음에는 날짜 정보의 안정성을 위해서  서버에서 처리하도록 요청하겠습니다) 

 날짜 선택 로직은  adapter에 람다를 통해 콜백함수를 전달해준뒤,   클릭 리스너에서 invoke하는 형식으로  viewModel에  선택된 날짜 정보를 넘겨서 처리하는 방향으로 진행하려 합니다.
